File: user-sites/user_60/brainfuck.c
#include <stdio.h>
#include <stdlib.h>

#define TAPE_SIZE 30000

void run(const char *code) {
    char tape[TAPE_SIZE] = {0};
    char *ptr = tape;
    const char *ip = code;
    
    while (*ip) {
        switch (*ip) {
            case '>': ptr++; break;
            case '<': ptr--; break;
            case '+': (*ptr)++; break;
            case '-': (*ptr)--; break;
            case '.': putchar(*ptr); break;
            case ',': *ptr = getchar(); break;
            case '[':
                if (*ptr == 0) {
                    int depth = 1;
                    ip++;
                    while (depth > 0) {
                        if (*ip == '[') depth++;
                        else if (*ip == ']') depth--;
                        ip++;
                    }
                    ip--;
                }
                break;
            case ']':
                if (*ptr != 0) {
                    int depth = 1;
                    ip--;
                    while (depth > 0) {
                        if (*ip == ']') depth++;
                        else if (*ip == '[') depth--;
                        ip--;
                    }
                    ip++;
                }
                break;
        }
        ip++;
    }
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <brainfuck code>\n", argv[0]);
        return 1;
    }
    run(argv[1]);
    return 0;
}

File: user-sites/user_60/brainfuck.py
#!/usr/bin/env python3
"""Brainfuck interpreter - functional test"""
import sys

def run(code):
    tape = [0] * 30000
    ptr = 0
    ip = 0
    code_len = len(code)
    
    while ip < code_len:
        cmd = code[ip]
        if cmd == '>':
            ptr = (ptr + 1) % 30000
        elif cmd == '<':
            ptr = (ptr - 1) % 30000
        elif cmd == '+':
            tape[ptr] = (tape[ptr] + 1) & 255
        elif cmd == '-':
            tape[ptr] = (tape[ptr] - 1) & 255
        elif cmd == '.':
            print(chr(tape[ptr]), end='', flush=True)
        elif cmd == ',':
            tape[ptr] = ord(sys.stdin.read(1)) & 255 if not sys.stdin.isatty() else 0
        elif cmd == '[':
            if tape[ptr] == 0:
                depth = 1
                ip += 1
                while depth > 0 and ip < code_len:
                    if code[ip] == '[':
                        depth += 1
                    elif code[ip] == ']':
                        depth -= 1
                    ip += 1
                ip -= 1
        elif cmd == ']':
            if tape[ptr] != 0:
                depth = 1
                ip -= 1
                while depth > 0 and ip >= 0:
                    if code[ip] == ']':
                        depth += 1
                    elif code[ip] == '[':
                        depth -= 1
                    ip -= 1
                ip += 1
        ip += 1

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <brainfuck_code>", file=sys.stderr)
        sys.exit(1)
    run(sys.argv[1])
